labs/multiverse/src/widgets/room_view/mod.rs

```diff
@@ -47,26 +74,124 @@ pub struct RoomView {
     current_pagination: Arc<Mutex<Option<JoinHandle<()>>>>,
 
     mode: Mode,
+    kind: TimelineKind,
 
-    timeline_list: ListState,
+    timeline_list: TimelineListState,
 
     input: Input,
 }
 
 impl RoomView {
-    pub fn new(ui_rooms: UiRooms, timelines: Timelines, status_handle: StatusHandle) -> Self {
+    pub fn new(client: Client, timelines: Timelines, status_handle: StatusHandle) -> Self {
         Self {
-            selected_room: None,
-            ui_rooms,
+            client,
             timelines,
             status_handle,
             current_pagination: Default::default(),
             mode: Mode::Normal { invited_room_view: None },
+            kind: TimelineKind::Room { room: None },
             input: Input::new(),
-            timeline_list: ListState::default(),
+            timeline_list: TimelineListState::default(),
         }
     }
 
+    fn switch_to_room_timeline(&mut self, room: Option<OwnedRoomId>) {
+        match &mut self.kind {
+            TimelineKind::Room { room: prev_room } => {
+                self.kind = TimelineKind::Room { room: room.or(prev_room.take()) };
+            }
+            TimelineKind::Thread { task, room, .. } => {
+                // If we were in a thread, abort the task.
+                task.abort();
+                self.kind = TimelineKind::Room { room: Some(room.clone()) };
+            }
+        }
+    }
+
+    fn switch_to_thread_timeline(&mut self) {
+        let Some(room) = self.room() else {
+            return;
+        };
+
+        let Some(timeline_list_nth) = self.timeline_list.selected() else {
+            return;
+        };
+
+        let Some(items) = self.get_selected_timeline_items() else {
+            self.status_handle.set_message("missing timeline for room".to_owned());
+            return;
+        };
+
+        let Some(root_event) = items.get(timeline_list_nth).and_then(|item| item.as_event()) else {
+            self.status_handle.set_message("no event associated to this timeline item".to_owned());
+            return;
+        };
+
+        if root_event.content().as_message().is_none() {
+            self.status_handle.set_message("this event can't be a thread start!".to_owned());
+            return;
+        }
+
+        let Some(root_event_id) = root_event.event_id().map(ToOwned::to_owned) else {
+            self.status_handle.set_message("can't open thread on a local echo".to_owned());
+            return;
+        };
+
+        self.status_handle.set_message(format!(
+            "Opening thread for event {root_event_id} in room {}",
+            room.room_id()
+        ));
```

Comment:
This status message is a bit visually displeasing, especially since it doesn't use the room name.

Unless you really need it for debugging purposes I would remove it.

Reply:
Yeah, I've kept it in the logs at least, because it helped show how long it takes to open a threaded view :+1: